File: scarl/src/main/scala/io/github/fiifoo/scarl/world/system/StellarBody.scala
package io.github.fiifoo.scarl.world.system

import io.github.fiifoo.scarl.world.system.source.StellarBodySourceId

object StellarBody {
  type Id = String
}

case class StellarBody(id: StellarBody.Id,
                       source: StellarBodySourceId,
                       mass: Double,
                       position: Position,
                       speed: Vector,
                      ) {

  def tick(tick: Int, bodies: Iterable[StellarBody]): StellarBody = {
    val acceleration = Physics.gravity(bodies, this)
    val speed = this.speed.add(acceleration.multiply(tick))
    val position = this.position.add(speed.multiply(tick))

    this.copy(speed = speed, position = position)
  }
}

File: scarl-ui/src/system/StellarBody.js
import { Record } from 'immutable'
import { gravity } from './physics'
import Position from './Position'
import Vector from './Vector'

const StellarBody = Record({
    id: undefined,
    source: undefined,
    mass: undefined,
    position: Position(),
    speed: Vector(),
})
StellarBody.read = ({position, speed, ...data}) => StellarBody({
    ...data,
    position: Position.read(position),
    speed: Vector.read(speed),
})
StellarBody.tick = (tick, bodies) => body => {
    const acceleration = gravity(bodies, body)

    const speed = Vector.add(body.speed, Vector.multiply(acceleration, tick))
    const position = Position.add(body.position, Vector.multiply(speed, tick))

    return body.set('speed', speed).set('position', position)
}

export default StellarBody
